doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+
+Whenever possible, it's best to reuse the evalf functionality defined in
+existing SymPy functions. However, in some cases it will be necessary to use
+mpmath directly.
+
+(custom-functions-rewriting-and-simplification)=
+### Rewriting and Simplification
+
+Various simplification functions and methods allow specifying their behavior
+on custom subclasses. Not every function in SymPy has such hooks. See the
+documentation of each individual function for details.
+
+(custom-functions-rewrite)=
+#### `rewrite()`
+
+The {meth}`~.rewrite` method allows rewriting an expression in terms of a
+specific function or rule. For example,
+
+```
+>>> sin(x).rewrite(cos)
+cos(x - pi/2)
+```
+
+To implement rewriting, define a method `_eval_rewrite(self, rule, args,
+**hints)`, where
+
+- `rule` is the *rule* passed to the `rewrite()` method. Typically `rule` will
+  be the class of the object to be rewritten to, although for more complex
+  rewrites, it can be anything. Each object that defines `_eval_rewrite()`
+  defines what rule(s) it supports. Many SymPy functions rewrite to common
+  classes, like `expr.rewrite(Add)`, to perform simplifications or other
+  computations.
+
+- `args` are the arguments of the function to be used for rewriting. This
+  should be used instead of `self.args` because any recursive expressions in
+  the args will be rewritten in `args` (assuming the caller used
+  `rewrite(deep=True)`, which is the default).
+
+- `**hints` are additional keyword arguments which may be used to specify the
+  behavior of the rewrite. Unknown hints should be ignored as they may be
+  passed to other `_eval_rewrite()` methods.
+
+
+The method should return a rewritten expression, using `args` as the
+arguments to the function, or `None` if the expression should be unchanged.
+
+For our [`versin` example](custom-functions-versine-definition), an obvious
+rewrite we can implement is rewriting `versin(x)` as `1 - cos(x)`:
+
+```py
+>>> class versin(Function):
+...     def _eval_rewrite(self, rule, args, **hints):
+...         if rule == cos:
+...             return 1 - cos(*args)
+>>> versin(x).rewrite(cos)
+1 - cos(x)
+```
+
+Once we've defined this, {func}`~.simplify` is now able to simplify some
+expressions containing `versin`:
+
+```
+>>> from sympy import simplify
+>>> simplify(versin(x) + cos(x))
+1
+```
+
+<!--
+
+TODO: simplify() can be customized with _eval_simplify(), but it isn't very
+powerful right now (see https://github.com/sympy/sympy/issues/19281). Most
+users are better off just defining rewrite(), doit(), and/or expand().
+
+#### Simplify
+
+-->
+
+(custom-functions-doit)=
+#### `doit()`
+
+The {meth}`doit() <sympy.core.basic.Basic.doit>` method is used to evaluate
+"unevaluated" functions. To define `doit()` implement `doit(self, deep=True,
+**hints)`. If `deep=True`, `doit()` should recursively call `doit()` on the
+arguments. `**hints` will be any other keyword arguments passed to the user,
```

Comment:
It would be much better to have `_eval_doit` so that subclasses don't need to implement recursion (#17280)

Reply:
Yes definitely. I noticed this problem for several methods. `expand` does it correctly, but most methods don't.